src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +616,278 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+				 * descended into sub join trees to handle outer joins buried in them.
+				 */
+				break;
+			}
+		}
+	}
+	else
+	{
+		ereport(ERROR, errmsg("got unexpected expr type when recursively "
+							  "planning a join"));
+	}
+}
+
+
+/*
+ * IsJoinNodeRecurring determines whether a join node in a join tree
+ * is recurring.
+ */
+static bool
+IsJoinNodeRecurring(Node *joinNode, Query *query)
+{
+	if (IsA(joinNode, JoinExpr))
+	{
+		/*
+		 * Then this is a sub (join) tree and here is how we determine whether a
+		 * join tree is recurring or not:
+		 *
+		 * i)   Outer join:
+		 *
+		 *      If the outer side of the join is a recurring rel, then the join
+		 *      node itself is recurring too regardless of whether the inner
+		 *      side of the join is recurring or not.
+		 *
+		 *      If outer side of the join is distributed, then the join
+		 *      node is distributed too.
+		 *
+		 * ii)  Inner join:
+		 *
+		 *      The join node is recurring only if both sides of the join are
+		 *      recurring.
+		 *
+		 *      Otherwise, if at least one side of the join is distributed, then
+		 *      this would ensure not returning recurring tuples from the workers;
+		 *      hence makes the join node "not recurring".
+		 */
+		JoinExpr *joinExpr = (JoinExpr *) joinNode;
+		bool leftNodeIsRecurring = IsJoinNodeRecurring(joinExpr->larg, query);
+		bool rightNodeIsRecurring = IsJoinNodeRecurring(joinExpr->rarg, query);
+
+		switch (joinExpr->jointype)
+		{
+			case JOIN_LEFT:
+			case JOIN_SEMI:
+			case JOIN_ANTI:
+			{
+				return leftNodeIsRecurring;
+			}
+
+			case JOIN_RIGHT:
+			{
+				return rightNodeIsRecurring;
+			}
+
+			case JOIN_FULL:
+			{
+				return leftNodeIsRecurring || rightNodeIsRecurring;
+			}
+
+			case JOIN_INNER:
+			{
+				return leftNodeIsRecurring && rightNodeIsRecurring;
+			}
+
+			default:
+			{
+				ereport(ERROR, (errmsg("unexpected join type")));
+			}
+		}
+	}
+	else if (IsA(joinNode, RangeTblRef))
+	{
+		RangeTblRef *rangeTableRef = (RangeTblRef *) joinNode;
+		int rangeTableIndex = rangeTableRef->rtindex;
+		List *rangeTableList = query->rtable;
+		RangeTblEntry *rangeTableEntry = rt_fetch(rangeTableIndex, rangeTableList);
+
+		/*
+		 * Inspect whether this side of the JOIN clause is recurring,
+		 * as in RelationInfoContainsOnlyRecurringTuples.
+		 */
+		return !FindNodeMatchingCheckFunctionInRangeTableList(list_make1(rangeTableEntry),
+															  IsDistributedTableRTE);
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("unexpected join node type")));
+	}
+}
+
+
+/*
+ * RecursivelyPlanNonRecurringJoinNode is a helper function for
+ * RecursivelyPlanRecurringLeftJoins that recursively plans given distributed
+ * node that is known to be inner side of an outer join.
+ */
+static void
+RecursivelyPlanNonRecurringJoinNode(Node *distributedNode, Query *query,
+									RecursivePlanningContext *recursivePlanningContext)
+{
+	if (IsA(distributedNode, JoinExpr))
+	{
+		/*
```

Comment:
Or, could we aggressively recursively plan all nodes below as much as possible? We do a similar hack in [`RecursivelyPlanAllSubqueries`](https://github.com/citusdata/citus/blob/80faf47ab5d7ad240c1cef2fa99d521f6ac1058e/src/backend/distributed/planner/recursive_planning.c#L671)

An example could be: https://github.com/citusdata/citus/commit/ca3c96280b4f247d4e153a624d7711c97969fd71

@marcocitus any concerns for doing that? 

Note: Most likely there is a hole in this commit, mostly for showing a possibility of doing it.

Reply:
no, I think that's ok / the only way to make this work within a limited timeframe